### PR TITLE
in_kafka: Fix in kafka json parsing

### LINKED ISSUE #7481: Kafka Input can't handle json data
## Bug Report

**Describe the bug**

@tarruda we just built an image with latest code (it includes kafka input), it can work well with plain text, but can't work well with json, please see below:

I defined Input Kafka, Output cloudwatch log group and stdout below is the configuration:

```
  inputs: |
    [INPUT]
        Name        kafka
        Brokers     $myKafkaServers
        Topics      $myKafkaTopic

        rdkafka.debug All
        rdkafka.enable.ssl.certificate.verification true
        rdkafka.ssl.certificate.location /certs/cert.pem
        rdkafka.ssl.key.location /certs/cert.key
        rdkafka.security.protocol ssl
        rdkafka.auto.offset.reset earliest
        rdkafka.enable.auto.commit false
        rdkafka.log_level 7

  outputs: |
    [OUTPUT]
        Name        cloudwatch_logs
        Match       *
        region us-east-1
        log_group_name /aws/clientLogTestGroup
        log_format json/emf
        log_stream_prefix fluent-bit-from-kafka-for-apic-
    [OUTPUT]
        Name stdout
        Match *

```
2. When I produce data to my kafkaTopic with plain text, I can get it from aws log group

$ ./kafka-console-producer.sh --broker-list $myKafkaServer --producer.config client.properties --topic $myKafkaTopic
`>test only`

I can get the text in aws log group:
image

if I produce json data, like below:
`>"{\"name\":\"jiajj\"}"`

or 
`>{"name":"jack"}`

It can't work, from log group, you can see the payload doesn't exist:

image

Would you please give some advice about it ? thanks !

**To Reproduce**
- Rubular link if applicable:
- Example log message if applicable:
```
{"log":"YOUR LOG MESSAGE HERE","stream":"stdout","time":"2018-06-11T14:37:30.681701731Z"}
```
- Steps to reproduce the problem:

**Expected behavior**

**Screenshots**

**Your Environment**

* Version used:
* Configuration:
* Environment name and version (e.g. Kubernetes? What version?):
* Server type and version:
* Operating System and version:
* Filters and plugins:

**Additional context**

### PR DESCRIPTION
**Pull request by tarruda**
This fixes json parsing for in_kafka, also add `data_format` configuration option, for explicit parsing control.

Fixes #7481

### CODE REVIEW

File: plugins/in_kafka/in_kafka.c [hunk 1]
```diff
@@ -246,6 +256,20 @@ static int in_kafka_init(struct flb_input_instance *ins,
         goto init_error;
     }
 
+    conf = flb_input_get_property("data_format", ins);
+    if (!conf) {
+        conf = "none";
+    }
+
+    if (!strcmp(conf, "none")) {
```
Comment (leonardo-albertovich) [inline plugins/in_kafka/in_kafka.c:264]: This has to be a caseless comparison and comparison against zero should be explicit

File: plugins/in_kafka/in_kafka.c [hunk 2]
```diff
@@ -128,10 +129,19 @@ static int process_message(struct flb_log_event_encoder *log_encoder,
 
     if (ret == FLB_EVENT_ENCODER_SUCCESS) {
         if (rkm->payload) {
-            if (try_json(log_encoder, rkm)) {
-                ret = flb_log_event_encoder_append_body_string(log_encoder,
-                                                               rkm->payload,
-                                                               rkm->len);
+            switch (ctx->data_format) {
```
Comment (leonardo-albertovich) [inline plugins/in_kafka/in_kafka.c:132]: This can be simplified to avoid repeating the code without adding the switch, please do so.

**Review by leonardo-albertovich — CHANGES_REQUESTED** [review]

File: plugins/in_kafka/in_kafka.c [hunk 3]
```diff
@@ -327,6 +345,11 @@ static struct flb_config_map config_map[] = {
     0, FLB_FALSE, 0,
     "Set the kafka topics, delimited by commas."
    },
+   {
+    FLB_CONFIG_MAP_STR, "data_format", (char *)NULL,
```
Comment (leonardo-albertovich) [inline plugins/in_kafka/in_kafka.c:349]: Set the default value for the new option here.

File: plugins/in_kafka/in_kafka.c [hunk 4]
```diff
@@ -246,6 +248,22 @@ static int in_kafka_init(struct flb_input_instance *ins,
         goto init_error;
     }
 
+    conf = flb_input_get_property("data_format", ins);
```
Comment (leonardo-albertovich) [inline plugins/in_kafka/in_kafka.c:251]: This configuration mechanism has been deprecated, add the fields offset to the config_map structure and use it directly.

**Review by leonardo-albertovich — CHANGES_REQUESTED** [review]
We should add this new setting to the documentation, especially if it changes the default behavior considering that before this plugin tried to decode json encoded data always (which is not the case anymore).

File: plugins/in_kafka/in_kafka.c (hunk 4 above)
Comment (tarruda) [reply plugins/in_kafka/in_kafka.c:251]: I'm aware of that, however it is not clear how we can handle "enum" properties using this mechanism. The data_format property currently supports 2 values, "none" and "json", but will likely support other values such as "avro" in the future.

Using the field offset, how can I check that the value set matches one of the valid strings (also ignoring case)?

Comment (leonardo-albertovich) [reply plugins/in_kafka/in_kafka.c:251]: We set it up as a string and then process that in the configuration function and set a different field with the numeric value.

**Review by leonardo-albertovich — DISMISSED** [review]

**Comment by patrick-stephens** [comment]
Looks like some repo failures so rekicked the unit tests off for the failing ones.

**Comment by edsiper** [comment]
All good with the changes; just one minor ask, please change the config property from `data_format` to `format` since we use it in other places.

**Comment by tarruda** [comment]
> All good with the changes; just one minor ask, please change the config property from `data_format` to `format` since we use it in other places.

renamed the property @edsiper

**Comment by lecaros** [comment]
@tarruda where's the doc for the new `format` option? What are the options for this setting?

**Comment by tarruda** [comment]
> @tarruda where's the doc for the new `format` option?

https://github.com/fluent/fluent-bit-docs/pull/1118  I just renamed `data_format` to `format` in the docs PR.

> What are the options for this setting?

For now it is just "none" (no parsing, the default) and "json". Eventually we might support other formats such as avro

**Comment by martinmajtan** [comment]
Hello,
I tried `format` option to my configuration and still I got this message:

`[2023/07/21 12:01:04] [error] [config] kafka: unknown configuration property 'format'. The following properties are allowed: poll_ms, topics, brokers, client_id, group_id, and rdkafka..`

I use this conf:
```

config:

  inputs: |
    [INPUT]
        Name kafka
        Brokers xxx
        Topics cip.webp.audit
        poll_ms 100
        format json
```

**Comment by tarruda** [comment]
@martinmajtan did you compile fluent-bit from latest master? I just ran the `kafka_filter` example and it is working for me.

**Comment by martinmajtan** [comment]
> @martinmajtan did you compile fluent-bit from latest master? I just ran the `kafka_filter` example and it is working for me.

I should have mention that, I'm using Helm chart where I used image version 2.1.7.

**Comment by tarruda** [comment]
@martinmajtan this PR will only be available in version 2.1.8

**Pull request merged**